csrc/scheduler/reduction_tma.cpp

```diff
@@ -7,25 +7,139 @@
+void scheduleReduction(Fusion* fusion, const TmaInnerReductionParams* pparams) {
   FusionGuard fg(fusion);
-  NVF_THROW("Schedule reduction using TMA");
+
+  // bool isUnrolled = rparams->isUnrolled();
+  bool isUnrolled = true;
+
+  // Cache inputs if unrolled
+  auto cached_inputs = scheduler_utils::cacheInputs(fusion, isUnrolled);
+
+  // Cache and fork outputs
+  auto cached_outputs =
+      scheduler_utils::cacheAndForkOutputs(fusion, isUnrolled);
+
+  // Make sure we don't have global memory set on intermediate tensors from
+  // fusion segmentation
+  scheduler_utils::clearMemorySpace(fusion);
+
+  scheduler_utils::prepareForMemoryTypePromotion(fusion);
+
+  [[maybe_unused]] std::vector<TensorView*> ldg_tvs, tma_tvs;
+  for (auto [tv, input_idx] : cached_inputs) {
+    if (auto load_op = dynamic_cast<LoadStoreOp*>(tv->definition())) {
+      load_op->setOpType(LoadStoreOpType::CpAsyncBulk);
+      tv->setMemoryType(MemoryType::Shared);
+      tma_tvs.push_back(tv);
+      // TODO cache to registers
+    } else {
+      ldg_tvs.push_back(tv);
+    }
+  }
+
+  NVF_ERROR(!tma_tvs.empty());
+
+  int64_t iter_axis = 0;
+  int64_t inner_reduce_axis = 1;
+
+  // Schedule TMA tvs as [BIDx, Bulk]
+  tma_tvs[0]->axis(iter_axis)->parallelize(ParallelType::BIDx);
+  tma_tvs[0]->axis(inner_reduce_axis)->parallelize(ParallelType::Bulk);
+  scheduler_utils::parallelizeAllLike(tma_tvs[0], tma_tvs);
+
+  // Schedule reduction tv
+  auto reduction_tvs = scheduler_utils::getReductionTvs(fusion);
+  NVF_ERROR(!reduction_tvs.empty());
+  TensorView* reduction_tv = reduction_tvs.at(0);
+
+  [[maybe_unused]] auto serialize = [&reduction_tv](
+                                        int64_t axis, int64_t factor) {
+    reduction_tv->split(axis, factor);
+    reduction_tv->axis(axis + 1)->parallelize(ParallelType::Serial);
+  };
+
+  [[maybe_unused]] auto vectorize = [&reduction_tv](
+                                        int64_t axis, int64_t factor) {
+    reduction_tv->split(axis, factor);
+    reduction_tv->axis(axis + 1)->parallelize(ParallelType::Vectorize);
+  };
+
+  [[maybe_unused]] auto inner_unswitch = [&reduction_tv](int64_t axis) {
+    reduction_tv->split(axis, 1);
+    reduction_tv->axis(axis + 1)->parallelize(ParallelType::Unswitch);
+  };
+
+  [[maybe_unused]] auto inner_unroll = [&reduction_tv](
+                                           int64_t axis, int64_t factor) {
+    reduction_tv->split(axis, factor);
+    reduction_tv->axis(axis + 1)->parallelize(ParallelType::Unroll);
+  };
+
+  [[maybe_unused]] auto inner_parallel_static =
+      [&reduction_tv](int64_t axis, ParallelType ptype, int64_t factor) {
+        reduction_tv->split(axis, factor);
+        reduction_tv->axis(axis + 1)->parallelize(ptype);
+      };
+
+  [[maybe_unused]] auto inner_parallel = [&reduction_tv](
+                                             int64_t axis, ParallelType ptype) {
+    reduction_tv->split(axis, NamedScalar::getParallelDim(ptype));
```

Comment:
What is the utility of this `inner_parallel()` which does not apply a static split factor? Is it choosing a split based on a TID parallel type? So it may choose a different value if this dim is split as TIDx vs TIDy?